### ISSUE TITLE
APEX inconsistent train_throughput in 0.8.0.dev6

### ISSUE THREAD

**Issue by carlbalmer** [ray-project/ray#6346]
# System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Debian GNU/Linux 9 
- **Ray installed from (source or binary)**: binary (nightly)
- **Ray version**: 0.8.0.dev6
- **Python version**: 3.6.3
- **Exact command to reproduce**: rllib train -f atari.yaml

### Describe the problem
On 0.8.0.dev6 Both train_throughput and sample_throughput are very erratic (jumping between 0 and 15k). On 0.7.6 they behave similar to the results published in [rl-experiments](https://github.com/ray-project/rl-experiments/tree/master/atari-apex/apex-results) (constant at ~2.5k).

But on both 0.8.0.dev6 and 0.7.6 the episode_reward_mean does not improve (still at ~2 at 4.5M steps)

![throughput](https://user-images.githubusercontent.com/6771285/70089787-2d67eb00-1619-11ea-86d0-4ad962a24632.png)
![reward](https://user-images.githubusercontent.com/6771285/70089833-44a6d880-1619-11ea-9034-64569da83c30.png)

The strange throughput values might just be a bug in the calculation of the metric, as apart from this the two versions behave similar.

The poor learning performance could be due to a lower train/sample steps ratio. Im getting only half the train throughput reported in the rl-experiments. Im running on similar hardware; 1 GPU 16CPUs per trial. What could be causing this? What parameters affect the train throughput?

### Source code / logs
[ray_results.zip](https://github.com/ray-project/ray/files/3918698/ray_results.zip)

atari.yaml:
```
apex:
    env: BreakoutNoFrameskip-v4
    run: APEX
    config:
        double_q: false
        dueling: false
        num_atoms: 1
        noisy: false
        n_step: 3
        lr: .0001
        adam_epsilon: .00015
        hiddens: [512]
        buffer_size: 1000000
        schedule_max_timesteps: 2000000
        exploration_final_eps: 0.01
        exploration_fraction: .1
        prioritized_replay_alpha: 0.5
        beta_annealing_fraction: 1.0
        final_prioritized_replay_beta: 1.0
        num_gpus: 1

        # APEX
        num_workers: 8
        num_envs_per_worker: 8
        sample_batch_size: 20
        train_batch_size: 512
        target_network_update_freq: 50000
        timesteps_per_iteration: 25000
```
cluster.yaml
```
cluster_name: gpu-apex-test

max_workers: 0

provider:
    type: gcp
    region: us-central1
    availability_zone: us-central1-a
    project_id: hanabi-ma-thesis

auth:
    ssh_user: ubuntu

head_node:
    machineType: n1-standard-16
    disks:
      - boot: true
        autoDelete: true
        type: PERSISTENT
        initializeParams:
          diskSizeGb: 50
          sourceImage: projects/deeplearning-platform-release/global/images/family/tf-1-13-cu100
    guestAccelerators:
      - acceleratorType: projects/hanabi-ma-thesis/zones/us-central1-a/acceleratorTypes/nvidia-tesla-p4
        acceleratorCount: 1
    metadata:
      items:
        - key: install-nvidia-driver
          value: "True"
    scheduling:
      - onHostMaintenance: TERMINATE

setup_commands:
    - >-
      wget https://repo.continuum.io/archive/Anaconda3-5.0.1-Linux-x86_64.sh -O ~/anaconda3.sh
      || true
      && bash ~/anaconda3.sh -b -p ~/anaconda3 || true
      && rm ~/anaconda3.sh
      && echo 'export PATH="$HOME/anaconda3/bin:$PATH"' >> ~/.profile

    - wget https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-0.8.0.dev6-cp36-cp36m-manylinux1_x86_64.whl
    # installing extras with whl-url did not work so im downloading it first
    - pip install -U ray-0.8.0.dev6-cp36-cp36m-manylinux1_x86_64.whl[debug,rllib] gputil /opt/deeplearning/binaries/tensorflow/tensorflow_gpu-1.13.1+nv-cp36-cp36m-linux_x86_64.whl
```

**Comment by ericl** [comment]
@edoakes do you see APEX breakout stalling at ~2 reward when running the compact-regression-test config on the release branch? APEX breakout should be hitting ~30 reward by 4.5m timesteps.

**Comment by carlbalmer** [comment]
My best guess is that the non improving reward is due tue the lower train throughput.
I tried it again with a v100-GPU and more CPUs (in case it was CPU bottle necking before); but got the same result (~2k train samples/sec, non improving reward).

Im not sure why the train throughput is lower or how to compensate for it... fewer workers? or smaller train batch size (in order to make more gradient steps)?

The Apex paper reports increased performance with more workers, but maybe this requires a minimum amount of train throughput. The paper reports ~9.7K train throughput, which is a lot higher than 2.5K...

**Comment by ericl** [comment]
Looks like we can confirm APEX is, once again, broken on master. Also, IMPALA seems to have suffered a 4x slowdown in throughput since b924299833c5ad47dbc331c51df5460d87d5e4e0 which is odd.
```
+----------------------------------------+------------+---------------------+--------+------------------+-------------+----------+
| Trial name                             | status     | loc                 |   iter |   total time (s) |   timesteps |   reward |
|----------------------------------------+------------+---------------------+--------+------------------+-------------+----------|
| IMPALA_BreakoutNoFrameskip-v4_78227f5c | TERMINATED |                     |    271 |          3601.1  |     1360000 |    28.4  |
| IMPALA_BreakoutNoFrameskip-v4_7822e6cc | TERMINATED |                     |    271 |          3600.91 |     1360500 |    22.46 |
| IMPALA_BreakoutNoFrameskip-v4_7823240c | TERMINATED |                     |    271 |          3603.04 |     1360500 |    23.62 |
| IMPALA_BreakoutNoFrameskip-v4_78235ff8 | TERMINATED |                     |    270 |          3601.99 |     1355000 |    42.43 |
| PPO_BreakoutNoFrameskip-v4_7823c3b2    | TERMINATED |                     |     30 |          3629.62 |      150000 |     2    |
| PPO_BreakoutNoFrameskip-v4_78242d84    | TERMINATED |                     |     30 |          3614.96 |      150000 |     1.93 |
| PPO_BreakoutNoFrameskip-v4_7824a2fa    | TERMINATED |                     |     30 |          3628.2  |      150000 |     2    |
| PPO_BreakoutNoFrameskip-v4_78250664    | TERMINATED |                     |     30 |          3620.32 |      150000 |     2.66 |
| APEX_BreakoutNoFrameskip-v4_78257f86   | TERMINATED |                     |    114 |          3623.5  |     4355360 |     2.45 |
| APEX_BreakoutNoFrameskip-v4_7825f1c8   | TERMINATED |                     |    114 |          3624.08 |     4284480 |     2.05 |
| APEX_BreakoutNoFrameskip-v4_78265d0c   | TERMINATED |                     |    114 |          3615.68 |     4332160 |     3.98 |
| APEX_BreakoutNoFrameskip-v4_7826cca6   | TERMINATED |                     |    114 |          3619.29 |     4336160 |     2.42 |
| A2C_BreakoutNoFrameskip-v4_782743b6    | RUNNING    | 172.31.29.115:26067 |    287 |          3010.6  |      865000 |    21.14 |
| A2C_BreakoutNoFrameskip-v4_7827a2f2    | RUNNING    | 172.31.29.115:26019 |    287 |          3012.1  |      865000 |    74.25 |
| A2C_BreakoutNoFrameskip-v4_7827fefa    | RUNNING    | 172.31.29.115:26040 |    285 |          3003.18 |      861000 |    30.2  |
| A2C_BreakoutNoFrameskip-v4_78285f9e    | RUNNING    | 172.31.29.115:26045 |    286 |          3001.53 |      878000 |    22.82 |
| DQN_BreakoutNoFrameskip-v4_7828c588    | PENDING    |                     |        |                  |             |          |
| DQN_BreakoutNoFrameskip-v4_7829537c    | PENDING    |                     |        |                  |             |          |
| DQN_BreakoutNoFrameskip-v4_7829c1a4    | PENDING    |                     |        |                  |             |          |
| DQN_BreakoutNoFrameskip-v4_782a2e96    | PENDING    |                     |        |                  |             |          |
+----------------------------------------+------------+---------------------+--------+------------------+-------------+----------+
```

**Comment by ericl** [comment]
One reason this might be happening is that TF is somehow not using GPUs. In that case sampling throughput will still be full speed, but training throughput will be greatly slowed down.

Are you seeing non-trivial GPU utilization?

Also, this patch will raise an error if this situation happens: https://github.com/ray-project/ray/pull/6365

**Comment by carlbalmer** [comment]
GPU utilization is at ~13% and ray_APEX:train() is listed as running process in nvidia-smi.

**Comment by carlbalmer** [comment]
@ericl Previously you posted the results of the compact-regression-test, but it seems to have been removed from the thread... was this a false alarm?

**Comment by edoakes** [comment]
That was indeed a false alarm. I was running the regression test as part of our release testing and had results inline with yours, but it ended up being caused by a misconfigured tensorflow environment. Here are the results running on an AWS g3.16xl instance with the deep learning AMI:

```
+----------------------------------------+------------+-------+--------+------------------+-------------+----------+
| Trial name                             | status     | loc   |   iter |   total time (s) |   timesteps |   reward |
|----------------------------------------+------------+-------+--------+------------------+-------------+----------|
| IMPALA_BreakoutNoFrameskip-v4_4445a400 | TERMINATED |       |    294 |          3600.28 |     6183500 |   352.15 |
| IMPALA_BreakoutNoFrameskip-v4_44461b92 | TERMINATED |       |    292 |          3610.1  |     6178500 |   147.14 |
| IMPALA_BreakoutNoFrameskip-v4_44467b28 | TERMINATED |       |    293 |          3611.05 |     6200500 |   275.1  |
| IMPALA_BreakoutNoFrameskip-v4_4446e1bc | TERMINATED |       |    294 |          3611.16 |     6161500 |   322.95 |
| PPO_BreakoutNoFrameskip-v4_44474a4e    | TERMINATED |       |    549 |          3603.09 |     2745000 |    36.41 |
| PPO_BreakoutNoFrameskip-v4_4447b740    | TERMINATED |       |    545 |          3604.43 |     2725000 |    16.83 |
| PPO_BreakoutNoFrameskip-v4_44481ea6    | TERMINATED |       |    549 |          3605.4  |     2745000 |    52.36 |
| PPO_BreakoutNoFrameskip-v4_444885a8    | TERMINATED |       |    546 |          3605.27 |     2730000 |    34.94 |
| APEX_BreakoutNoFrameskip-v4_4448feac   | TERMINATED |       |    113 |          3629.93 |     3552960 |    13.47 |
| APEX_BreakoutNoFrameskip-v4_44497684   | TERMINATED |       |    112 |          3615.5  |     3539360 |    28.65 |
| APEX_BreakoutNoFrameskip-v4_4449f01e   | TERMINATED |       |    113 |          3621.85 |     3524640 |    28.17 |
| APEX_BreakoutNoFrameskip-v4_444a5dd8   | TERMINATED |       |    112 |          3616.41 |     3447520 |    20.56 |
| A2C_BreakoutNoFrameskip-v4_444acb7e    | TERMINATED |       |    349 |          3604.38 |     2981000 |   108.8  |
| A2C_BreakoutNoFrameskip-v4_444b3a50    | TERMINATED |       |    349 |          3602.3  |     2967500 |   145.82 |
| A2C_BreakoutNoFrameskip-v4_444b9ebe    | TERMINATED |       |    349 |          3605.71 |     2990500 |   127.3  |
| A2C_BreakoutNoFrameskip-v4_444c0016    | TERMINATED |       |    349 |          3602.58 |     2970000 |   112.92 |
| DQN_BreakoutNoFrameskip-v4_444c6830    | TERMINATED |       |     30 |          3662.81 |      300000 |    18.71 |
| DQN_BreakoutNoFrameskip-v4_444cdeb4    | TERMINATED |       |     30 |          3608.95 |      300000 |    17.45 |
| DQN_BreakoutNoFrameskip-v4_444d5038    | TERMINATED |       |     31 |          3658.65 |      310000 |    17.43 |
| DQN_BreakoutNoFrameskip-v4_444dbfb4    | TERMINATED |       |     31 |          3696.8  |      310000 |    18.63 |
+----------------------------------------+------------+-------+--------+------------------+-------------+----------+
```

**Comment by carlbalmer** [comment]
OK. Thank you for your help. I think the behavior I saw is due to running on different hardware (with different sample and train throughput) and could be solved by tuning the hyper parameters.
closing the issue...
